### PR TITLE
Switch to synchronous I/O (for GPIO and I2C)

### PR DESCRIPTION
**Pull request by sandeepmistry**
Might help with #15 and #17 - needs testing.

Can be tested via: `npm install sandeepmistry/node-chip-io#sync-io`

cc/ @nkolban @rwaldron @ugate @sumarouno @ericterpstra

### CODE REVIEW

**Comment by sumarouno** [comment]
Hi @sandeepmistry , thanks you for mentioning me. I tried your PR. 
`npm install sandeepmistry/node-chip-io#sync-io`

And then, try previous code #15 and try some code for my lcd 1602 with pcf8374T i2c backpack. Here the code I have tried:
lcd_hello_world.js:
```nodejs
"use strict";

var five = require('johnny-five');
var chipio = require('chip-io');

var board = new five.Board({
	io: new chipio()
});

board.on('ready', function() {
	let lcd = new five.LCD({
		controller: "PCF8574A", 
		address: 0x27, 
		bus: 2
	});
	lcd.cursor(0,0);
	lcd.print("Hello World");
	lcd.cursor(1,0);
	lcd.print("1234567890");
	console.log("Percobaan hello world!");
});
```
lcd_custom_char.js:
```javascript
var five = require("johnny-five");
var chipio = require('chip-io');

var board = new five.Board({
  io: new chipio()
});


board.on("ready", function() {
  var l = new five.LCD({
        controller: "PCF8574T",
        address: 0x27,
        bus: 2,
        rows: 2,
        cols: 16
  });
	l.useChar("heart");
	l.cursor(0, 0).print("hello, I :heart: JS");
	l.cursor(1, 0).print("NodeJs is great");

});
```
All above code work well after using sync-io. Run the code, stop the code, run again, stop again. And no more garbage text as discussed #15 .
I can confirm, no more garbage text again in my lcd.
Your work is great, thank you very much.

**Comment by sandeepmistry** [comment]
@sumarouno excellent thanks for trying out the changes!

I'll hold off merging for a week or so, so most testing can be done.

**Comment by sumarouno** [comment]
I will try more code as soon as possible. 
Fyi, this PR still not solve for #17

File: lib/pcf8574a.js [hunk 1]
```diff
@@ -54,32 +50,26 @@ PCF8574A.prototype.digitalRead = function(pin, value) {
...
+    var value = this._read();
+
+    for (var i = 0; i < 8; i++) {
+      var pinMask = (1 << i);
+
+      if (this._readMask & pinMask) {
+        this.emit('digital-read', 'XIO-P' + i, (value & pinMask) ? 1 : 0);
       }
-    }.bind(this));
+    }
```
Comment (rwaldron) [inline lib/pcf8574a.js]: Digital reading should be async

File: lib/axp209.js [hunk 2]
```diff
@@ -40,68 +39,45 @@ AXP209.prototype.digitalWrite = function(pin, value) {
...
-
-    this._i2c.readRegister(lsbRegister, 1, function(err, lsbData) {
-      if (err) {
-        return callback(err);
-      }
-
-      var value = (msbData[0] << 4) | (lsbData[0] & 0x0f);
+AXP209.prototype._readAdc = function(msbRegister, lsbRegister) {
+  msbData = this._i2c.readRegister(msbRegister, 1);
+  lsbData = this._i2c.readRegister(lsbRegister, 1);
```
Comment (rwaldron) [inline lib/axp209.js]: Analog reading needs to be async

**Comment by rwaldron** [comment]
- Reading needs to be async
- Writing can be async, but only if order can be guaranteed and there is no dependency on a callback before the next operation—otherwise just make the write sync (generally this is faster anyway)

File: lib/pcf8574a.js (hunk 1 above)
Comment (sandeepmistry) [inline lib/pcf8574a.js]: We're good here, because all reads are done on the timer tick.

File: lib/axp209.js (hunk 2 above)
Comment (sandeepmistry) [inline lib/axp209.js]: Same as digital read comment, all reads are done on the tick timer.

File: lib/chip-io.js [hunk 3]
```diff
@@ -253,42 +242,24 @@ ChipIO.prototype.i2cReadOnce = function(address, register, size, handler) {
...
-      i2c.read(size, function(err, data) {
-        if (err) {
-          return;
-        }
+  if (register) {
+    event = 'I2C-reply-' + address + '-' + register;
+    data = i2c.readRegister(register, size);
+  } else {
+    event = 'I2C-reply-' + address;
+    data = i2c.read(size);
```
Comment (sandeepmistry) [inline lib/chip-io.js]: I've switched it to use `process.nextTick` for now. We can look at it again when we switch to use `i2s-bus`.

**Comment by sandeepmistry** [comment]
@rwaldron thanks for reviewing and providing your feedback. I think everything is up to spec now.

**Comment by rwaldron** [comment]
@sandeepmistry you're the best!

**Pull request merged**